File: fsharp/rectangles/Rectangles.fs
module Rectangles

let rectangles lines = 
    let get_corner_index_pairs line = 
        (Seq.allPairs line line)
        |> Seq.filter(fun (a, b) -> a < b)

    let corner_index_pairs = 
        lines 
        |> Seq.map(fun line -> 
            line 
            |> Seq.indexed
            |> Seq.filter(fun (_, c) -> c = '+') 
            |> Seq.map(fun (i, _) -> i) 
            |> get_corner_index_pairs)
        |> Seq.indexed

    let get_row2_candidates pair from_row =
        corner_index_pairs
        |> Seq.skip (from_row + 1)
        |> Seq.filter(fun (_, pairs) -> pairs |> Seq.contains pair)
        |> Seq.map(fun (row2, _) -> row2)

    let horizontal = ['+'; '-']
    let vertical = ['+'; '|']

    let check_rectangle row1 row2 (col1, col2) =
        lines                  |> Seq.item row1 |> Seq.skip (col1 + 1) |> Seq.take (col2 - col1 - 1) |> Seq.forall (fun c -> horizontal |> Seq.contains c) &&
        lines                  |> Seq.item row2 |> Seq.skip (col1 + 1) |> Seq.take (col2 - col1 - 1) |> Seq.forall (fun c -> horizontal |> Seq.contains c) &&
        lines |> Seq.transpose |> Seq.item col1 |> Seq.skip (row1 + 1) |> Seq.take (row2 - row1 - 1) |> Seq.forall (fun c -> vertical   |> Seq.contains c) &&
        lines |> Seq.transpose |> Seq.item col2 |> Seq.skip (row1 + 1) |> Seq.take (row2 - row1 - 1) |> Seq.forall (fun c -> vertical   |> Seq.contains c)

    corner_index_pairs 
    |> Seq.map(fun (row1, pairs) -> 
        pairs 
        |> Seq.map(fun pair ->
            (get_row2_candidates pair row1)
            |> Seq.filter(fun row2 -> check_rectangle row1 row2 pair)
            |> Seq.length
            )
        |> Seq.sum
        )
    |> Seq.sum

File: csharp/rectangles/Rectangles.cs
using System;
using System.Linq;
using System.Collections.Generic;

public static class Rectangles
{
    public static int Count(string[] rows)
    {
        if (rows.Length == 0 || rows.Where(row => row.Contains('+')).Count() == 0) return 0;
        var cornerPairs = GetCornerPairs();
        var transpose = TransposeInput();
        var horizontal = new char[] { '+', '-' };
        var vertical = new char[] { '+', '|' };

        return cornerPairs
            .Select((pairs, row1) =>
                pairs.Select(pair => CountRectanglesForPair(pair, row1)).Sum())
            .Sum();

        // Local functions

        List<List<(int, int)>> GetCornerPairs()
        {
            var result = new List<List<(int, int)>>();
            foreach (string row in rows)
            {
                var indices = row.Select((c, i) => (c, i)).Where(el => el.c == '+').Select(el => el.i).ToArray();
                var cornerPairs = new List<(int, int)>();
                for (var i = 0; i < indices.Length - 1; i++)
                {
                    for (var j = i + 1; j < indices.Length; j++)
                    {
                        cornerPairs.Add((indices[i], indices[j]));
                    }
                }
                result.Add(cornerPairs);
            }
            return result;
        }
        
        string[] TransposeInput()
        {
            var transpose = new string[rows[0].Length];
            for (var i = 0; i < transpose.Length; i++)
            {
                var line = new string("");
                for (var j = 0; j < rows.Length; j++)
                {
                    line += rows[j][i];
                }
                transpose[i] = line;
            }
            return transpose;
        }

        int CountRectanglesForPair((int, int) pair, int row1)
        {
            return cornerPairs
                .Select((pairs, row2) => (pairs, row2))
                .Skip(row1 + 1)
                .Where(el => el.pairs.Contains(pair) && CheckRectangle(row1, el.row2, pair))
                .Count();
        }

        bool CheckRectangle(int row1, int row2, (int, int) columns)
        {
            int col1 = columns.Item1;
            int col2 = columns.Item2;
            return rows[row1].Substring(col1 + 1, col2 - col1 - 1).All(c => horizontal.Contains(c))
                && rows[row2].Substring(col1 + 1, col2 - col1 - 1).All(c => horizontal.Contains(c))
                && transpose[col1].Substring(row1 + 1, row2 - row1 - 1).All(c => vertical.Contains(c))
                && transpose[col2].Substring(row1 + 1, row2 - row1 - 1).All(c => vertical.Contains(c));
        }
    }
}